### PR TITLE
split encoding.h

### PR DESCRIPTION
**Pull request by shyouhei**
After I organise header files, this is the largest one now.  I think it's a bit too large to read through.  Why not split it into files.

### CODE REVIEW

File: string.c [hunk 1]
```diff
@@ -6891,6 +6891,7 @@ upcase_single(VALUE str)
 
     while (s < send) {
         unsigned int c = *(unsigned char*)s;
+        const rb_encoding *const enc = 0;
```
Comment (nobu) [inline string.c:6894]: Rather the call of `rb_enc_isascii` isn't unnecessary?
`'a' <= c && c <= 'z'` implies "isascii" enough, I guess.

File: string.c [hunk 2]
```diff
@@ -5732,7 +5732,7 @@ rb_str_setbyte(VALUE str, VALUE index, VALUE value)
 		ENC_CODERANGE_SET(str, ENC_CODERANGE_VALID);
 	    goto end;
 	  case ENC_CODERANGE_VALID:
-	    left = rb_enc_left_char_head(head, ptr, head+len, enc);
+            left = rb_enc_left_char_head(head, (char *)ptr, head+len, enc);
```
Comment (nobu) [inline string.c:5735]: Shouldn't `ptr` be a plain `char` pointer?

File: string.c (hunk 1 above)
Comment (shyouhei) [reply string.c:6894]: Sounds reasonable.

File: string.c (hunk 2 above)
Comment (shyouhei) [reply string.c:5735]: That works _for me_, but could break this line in some environment:

https://github.com/ruby/ruby/blob/3978a2abb86ae85325bdc8fedb9bf08ed0d65248/string.c#L5737

If `ptr` is plain `char`, and if the environment chooses it's signed, then unsigned to signed conversion is allowed to fail.

Comment (nobu) [reply string.c:5735]: It may need a cast, but I think that casting signedness is preferable to pointer punning.

Comment (nobu) [reply string.c:5735]: Well, this case is not "punning", still I prefer value-casts to pointer-casts.

**Comment by nobu** [comment]
I meant
```diff
diff --git a/string.c b/string.c
index 78e2ba923f3..7c2235450d6 100644
--- a/string.c
+++ b/string.c
@@ -5706,7 +5706,7 @@ rb_str_setbyte(VALUE str, VALUE index, VALUE value)
     long pos = NUM2LONG(index);
     long len = RSTRING_LEN(str);
     char *head, *left = 0;
-    unsigned char *ptr;
+    char *ptr;
     rb_encoding *enc;
     int cr = ENC_CODERANGE_UNKNOWN, width, nlen;
 
@@ -5717,18 +5717,18 @@ rb_str_setbyte(VALUE str, VALUE index, VALUE value)
 
     VALUE v = rb_to_int(value);
     VALUE w = rb_int_and(v, INT2FIX(0xff));
-    unsigned char byte = NUM2INT(w) & 0xFF;
+    char byte = (char)(NUM2INT(w) & 0xFF);
 
     if (!str_independent(str))
 	str_make_independent(str);
     enc = STR_ENC_GET(str);
     head = RSTRING_PTR(str);
-    ptr = (unsigned char *)&head[pos];
+    ptr = &head[pos];
     if (!STR_EMBED_P(str)) {
 	cr = ENC_CODERANGE(str);
 	switch (cr) {
 	  case ENC_CODERANGE_7BIT:
-            left = (char *)ptr;
+            left = ptr;
 	    *ptr = byte;
 	    if (ISASCII(byte)) goto end;
 	    nlen = rb_enc_precise_mbclen(left, head+len, enc);
```

**Comment by shyouhei** [comment]
@nobu That's also okay.  But ultimately, `ONIGENC_LEFT_ADJUST_CHAR_HEAD` operates over `OnigUChar*`.  Using `unsigned char` here seems a bit more natural, to me.

**Comment by nobu** [comment]
But `rb_enc_left_char_head` does not use `unsigned char`.

**Pull request merged**